lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,269 @@
+        """
+        Return a hashed version of the list of block records.
+
+        This currently hashes using sha256, and returns a base64 encoded version
+        of the binary digest.  In the future, different algorithms could be
+        supported by adding a label indicated which algorithm was used, e.g.,
+        "sha1$witfkXg0JglCjW9RssWvTAveakI=".
+        """
+        return b64encode(sha256(self.to_json()).digest())
+
+
+class VisibleBlocksQuerySet(models.QuerySet):
+    """
+    A custom QuerySet representing VisibleBlocks.
+    """
+    def create_from_blockrecords(self, blocks):
+        """
+        Creates a new VisibleBlocks model object.
+
+        Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        if not isinstance(blocks, BlockRecordList):
+            blocks = BlockRecordList(blocks)
+        model, _ = self.get_or_create(hashed=blocks.to_hash(), defaults={'blocks_json': blocks.to_json()})
+        return model
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a
+    given subsection at the time they are used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored
+    in the blocks_json field. A hash of this json array is used for lookup
+    purposes.
+    """
+    blocks_json = models.TextField()
+    hashed = models.CharField(max_length=44, unique=True)
+
+    objects = VisibleBlocksQuerySet.as_manager()
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return "VisibleBlocks object - hash:{}, raw json:'{}'".format(self.hashed, self.blocks_json)
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as a list of
+        BlockRecords in the order they were provided.
+        """
+        block_dicts = json.loads(self.blocks_json)
+        record_generator = (BlockRecord(data['locator'], data['weight'], data['max_score']) for data in block_dicts)
+        return BlockRecordList(record_generator)
+
+
+class PersistentSubsectionGradeQuerySet(models.QuerySet):
+    """
+    A custom QuerySet, that handles creating a VisibleBlocks model on creation, and
+    extracts the course id from the provided usage_key.
+    """
+    def create(self, **kwargs):
+        """
+        Instantiates a new model instance after creating a VisibleBlocks instance.
+
+        Arguments:
+            user_id (int)
+            usage_key (serialized UsageKey)
+            course_version (str)
+            subtree_edited_date (datetime)
+            earned_all (float)
+            possible_all (float)
+            earned_graded (float)
+            possible_graded (float)
+            visible_blocks (iterable of BlockRecord)
+        """
+        visible_blocks = kwargs.pop('visible_blocks')
+        if not isinstance(visible_blocks, BlockRecordList):
```

Comment:
Same question - when would `visible_blocks` not be a `BlockRecordList`?  Or will it always not be?